File: 131.分割回文串.py
#
# @lc app=leetcode.cn id=131 lang=python3
#
# [131] 分割回文串
#

# @lc code=start
from typing import List
class Solution:
    def func(self, mask: int, s: str = None):
        if s is None: s = self.s
        idx = 1
        ret, rret = [0], []
        while mask:
            if mask & 1: ret.append(idx)
            mask >>= 1
            idx += 1
        ret.append(len(s))
        for i in range(1, len(ret)):
            rret.append(s[ret[i-1]:ret[i]])
            if rret[-1] != rret[-1][::-1]: return None
        return rret
        
    def partition(self, s: str) -> List[List[str]]:
        self.s = s
        ret = []
        for mask in range(1<<(len(s) - 1)):
            tmp = self.func(mask)
            if tmp is not None:
                ret.append(tmp)
        return ret

# @lc code=end
print(Solution().partition("0123456789"))
print(Solution().partition("aab"))
print(Solution().partition("a"))
print(Solution().partition("a" * 16))

File: 131.分割回文串.cpp
/*
 * @lc app=leetcode.cn id=131 lang=cpp
 *
 * [131] 分割回文串
 */

// @lc code=start
#include<vector>
#include<optional>
#include<string>
#include<iostream>
using namespace std;
class Solution {
public:
    optional<vector<string>> func(int mask, const string& s){
        int idx = 1;
        vector<int> idxs = {0};
        vector<string> ret;
        while(mask){
            if(mask & 1) idxs.push_back(idx);
            mask >>= 1;
            ++idx;
        }
        idxs.push_back(s.size());
        for(int i=1; i<idxs.size(); ++i){
            string tmp = s.substr(idxs[i-1], idxs[i] - idxs[i-1]);
            ret.push_back(tmp);
            reverse(tmp.begin(), tmp.end());
            if(ret.back() != tmp) {
                return nullopt;
            }
        }
        return ret;
    }
    vector<vector<string>> partition(string s) {
        vector<vector<string>> ret;
        for(int mask = 0; mask < (1<<(s.size()-1)); ++mask){
            auto tmp = func(mask, s);
            if(tmp.has_value()) {
                ret.emplace_back(tmp.value());
            }
        }
        return ret;
    }
};
// @lc code=end

